### ISSUE TITLE
Certbot for Windows is not using ProgramData (the equivalent of /etc on Linux)

### ISSUE THREAD

**Issue by oxygen** [certbot/certbot#7872]
Certbot for Windows is not using ProgramData (the equivalent of /etc on Linux).

## My operating system is (include version):
Any Windows version

## I installed Certbot with (certbot-auto, OS package manager, pip, etc):
Certbot for Windows Installer beta

## I ran this command and it produced this output:
certbot certonly -d somedomain.com -d somedomain.com

## Certbot's behavior differed from what I expected because:
Certbot is placing the generated files (certificates and such) in C:\Certbot\ **instead of** C:\ProgramData\Certbot

**Comment by adferrand** [comment]
Hello @oxygen!

While porting Certbot on Windows, we hesitated between several path for the equivalent of `/etc/letsencrypt`, and ended to simply select `C:\Certbot`. We evicted `C:\ProgramData` for some of the following reasons, that can also apply to other potential locations.

First, there is not strict equivalent of `/etc`. Over the years Microsoft added several default paths, but none of them became a de facto standard for storing the configuration. Some programs stores their config in their installation path, other in `AppData` when the program is limited to one user, others in a Posix user style like `$HOME/.oneconfig` and so on. Very few in facts use `C:\ProgramData`, and most of the time is just for caching purpose that are not specific to one user

But the biggest usage is to store the config in the registry, because Windows is does not follow the "everything is a file" pattern like in Posix systems.

Also generally it is not expected to have critical data in `C:\ProgramData`, and this folder is hidden

For all these reasons, we preferred to go to the simplest path for the general purpose.

**Comment by oxygen** [comment]
**/etc has a single equivalent on Windows**, %ProgramData%, normally C:\ProgramData if C: is your system drive.

For processes running under user accounts, all the directory paths under /Users are users, and each is just a home folder.

The Windows Registry (made up of multiple hives, each mounted in HKLM, HKCU (from home folders) and some subpaths, such as SAM and BCD store) has an equivalent on all OS:
 - On linux it is more or less configfs
 - On OS X the .plist files, where each .plist file is exactly like a Windows Registry hive in functionality (both binary and text representations).

Keep in mind Windows has a much more advanced security model and ProgramData is specifically authorized for access by processes running under [LocalSystem Account](https://docs.microsoft.com/en-us/windows/win32/services/localsystem-account), [NetworkService Account](https://docs.microsoft.com/en-us/windows/win32/services/networkservice-account), and so on.

There is no guarantee for files placed right under C:\ to be accessible. Worse, in special Windows setups Certbot is likely to fail as the C:\Certbot path might fall into a read-only path (while ProgramData would be a mount point for service data and configuration).

System services definately need to be able to read the Certbot files and that is guaranteed in Windows only for the C:\ProgramData path. Also, path writability as a system service is only guaranted for c:\ProgramData paths, while Program Files and others are not.

System services will definately not want to read files from C:\Users subpaths.
Normally, a computer admin has read-only access to C:\ProgramData by default.

In short:
 - There is only one /etc equivalent on Windows, and that is C:\ProgramData
 - Writing at the current C:\Certbot path presents risks for both present Windows users and worse, future Windows versions with stricter security.

All Windows software equivalents (which are prevalent in Linux) that I know of, and which run as system services (the kind of programs Certbot is assisting) load and write data and configuration from C:\ProgramData (and most do it correctly by getting the path from the environment variables).

**Comment by oxygen** [comment]
To avoid unnecessary pain in the future, please conform to what Windows has standardized as the right path for configuration.

**Comment by adferrand** [comment]
I am really sorry, but there is no standard here. The official documentation is quite vague about its usage (https://docs.microsoft.com/en-us/windows/win32/shell/csidl?redirectedfrom=MSDN or https://docs.microsoft.com/en-us/windows-hardware/customize/desktop/unattend/microsoft-windows-shell-setup-folderlocations-programdata).

And for what it describes, it does not apply to Certbot. Indeed it states that content should be not user specific, while a significant part of content in Certbot config folder must have very restrictive accesses (typically the account key and the certificate private keys).

I am conscious that the model used in Certbot is still very tied to Unix philosophy, and several things could be improved here to be more Windows native. You mentionned hardlinks vs symlinks, there is also the usage of the Certificate store, that is specifically designed for this usage, and would improve a lot the future interoperability for IIS. There is no proper way also to store securely files on a CIFS mount and interoperability with Active Directory.

All these things will be implemented eventually, but it will take time considering the limited capacity we have in the team. If you are willing to accelerate things, you can also participate by submitting PRs, I will be glad to review them.

For what matter in this specific issue, I am still convinced that `ProgramData` is not the good place to store Certbot data. However, I think also that Certbot should be quite enough flexible to match different situations: here it is true that the only effective default config path you can use is `C:\Certbot` without modifying some internals of Certbot, in particular the scheduled tasks for renewal.

One first step here will be to give to Windows installer the capacity to define in the registry what is the effective default config path to use for a given system. This should be an option that would be given in the installer UI, very similarly to the installation path.

**Comment by oxygen** [comment]
You have made several untrue statements.

"Over the years Microsoft added several default paths" - Nope.

"Some programs store their config in their installation path" - Nope. Maybe 20 years ago in Windows XP and Windows 98.

"other in AppData" - AppData us a subfolder of user home folders, which are the exact equivalent of user home directories on Linux. Not /etc. 

"But the biggest usage is to store the config in the registry, because Windows is does not follow the "everything is a file" pattern like in Posix systems." - Of course. It is a hierarchycal database with strict data types for fast memory access, primarily from the kernel to the SAM, BCD and SYSTEM registry hives (direct equivalent of .plist and similar to configfs when having to share configuration with the kernel).

"Also generally it is not expected to have critical data in C:\ProgramData" - Yes it is, several big name databases use it.

/etc is NOT a user homefolder (keeps configuration)
/var/lib is NOT a user homefolder (keeps data)

/ProgramData (keeps data and configuration).

That's that.
Putting your data and configuration under the root mountpoint (/ or generally known as c:\) is not acceptable in any OS.

The dangers for Windows I have enumerated in the previous post.

Certbot will be a big hit and will survive very long years. 
You absolutely need to get started right with the paths or you will cause great pain to many people.

**Comment by adferrand** [comment]
> 
> 
> You have made several untrue statements.
> 
> "Over the years Microsoft added several default paths" - Nope.

https://docs.microsoft.com/en-us/windows/win32/shell/csidl?redirectedfrom=MSDN
60+ entries

> 
> "Some programs store their config in their installation path" - Nope. Maybe 20 years ago in Windows XP and Windows 98.

https://gitforwindows.org/

> 
> "other in AppData" - AppData us a subfolder of user home folders, which are the exact equivalent of user home directories on Linux. Not /etc.

Again, `ProgramData` is not `/etc`. It is `C:\Users\All Users`, so more like a user folder for everyone. Try hitting `C:\Users\All Users` in your Windows explorer.

> 
> "But the biggest usage is to store the config in the registry, because Windows is does not follow the "everything is a file" pattern like in Posix systems." - Of course. It is a hierarchycal database with strict data types for fast memory access, primarily from the kernel to the SAM, BCD and SYSTEM registry hives (direct equivalent of .plist and similar to configfs when having to share configuration with the kernel).

And again I think we will store the default configuration folder for the current Certbot installation of a given machine in it. It is quite the place to put that.

> 
> "Also generally it is not expected to have critical data in C:\ProgramData" - Yes it is, several big name databases use it.

So again, not similar to `/etc`. On Linux database usually store data in `/var/lib`.

> 
> /etc is NOT a user homefolder (keeps configuration)
> /var/lib is NOT a user homefolder (keeps data)
> 
> /ProgramData (keeps data and configuration).

So, given your example on databases, or the equivalency to `C:\Users\All Users` as I said above, `ProgramData` will be more a place for data, not configuration. But anyway, I put again the official description from Microsoft here: https://docs.microsoft.com/en-us/windows-hardware/customize/desktop/unattend/microsoft-windows-shell-setup-folderlocations-programdata

> 
> That's that.
> Putting your data and configuration under the root mountpoint (/ or generally known as c:) is not acceptable in any OS.

There is no concept of root mountpoint in Windows, because it is not file centric, so it does not need a root described as a filesystem. There are only mount points, where drive letters are the most usual, and `C:` happens to be among them the traditional letter chosen by Windows installation process for the first drive.

Which leads me again to the problematic that there is no strong convention on file systems on Windows, and I think we will need to allow users to define their own default configuration path if they want during the installation phase.

> 
> The dangers for Windows I have enumerated in the previous post.
> 
> Certbot will be a big hit and will survive very long years.
> You absolutely need to get started right with the paths or you will cause great pain to many people.

There is no need for that dramatic assertion, we are simply discussing about technical specifications here. `C:\ProgramData` still seems a wrong location in my opinion, and `C:\Certbot` the least wrong for a default choice, suitable if it is overriddable during the installation phase.

**Comment by oxygen** [comment]
Non-elevated process in Windows can read ProgramData and, **lo and behold**, can't read c:\Certbot\live

```
[17452] Unhandled Rejection at: Promise Promise {
  <rejected> { Error: EPERM: operation not permitted, stat 'C:\Certbot\live\README'
    errno: -4048,
    code: 'EPERM',
    syscall: 'stat',
    path: 'C:\\Certbot\\live\\README' } } reason { [Error: EPERM: operation not permitted, stat 'C:\Certbot\live\README']
  errno: -4048,
  code: 'EPERM',
  syscall: 'stat',
  path: 'C:\\Certbot\\live\\README' }
```

**Comment by adferrand** [comment]
OK we starting to have something here.

There is a thing to improve before the main topic about configuration location you are describing here. For now, Certbot requires administrative privileges to run. It does so because symlinks need to be created to feed the live directory. This behavior is imported from the Unix world. However Windows requires (stupidly?) administrative privileges to create symlinks.

This is why most content of `C:\Certbot` is not accessible to normal users, and it would be the same for `C:\ProgramData\Certbot`.

We need to fix that, and I opened #7880 for it. Once done, we should modify Certbot to let it run without administrative privileges (related issue #7881). Finally, we can modify the Windows installer to offer Certbot installation as user oriented or system oriented (administrative privileges for the latter): at this point, the default decent location path for a user install could be `APPDATA`. And also in that context, the default decent location for a system install could be `PROGRAMDATA`. This is described in #7883, with maybe a global solution for default folder locations described in #7882.
